### PR TITLE
BUG: Fix crackfortran groups for endifs with comments

### LINKED ISSUE #23533: BUG: f2py.crackfortran fails on comment after endif statement
### Describe the issue:

Hey there!

`numpy.f2py.crackfortran.crackfortran` fails, when there is a comment after an `END <IF|DO|etc.>` statement. I'm guessing that this is due to [the `endifpattern` here](https://github.com/numpy/numpy/blob/c16a05ce3877bed28b9880829c169271051ef734/numpy/f2py/crackfortran.py#LL619C13-L619C13) only checking for `r'[\w\s]*'` after the statement, while comments start with a `!`. In contrast to this, `endpattern` a few lines above uses the more generalized `r'.*'`.

### Reproduce the code example:

```python
import tempfile
from numpy.f2py import crackfortran

FORTRAN = """\
    SUBROUTINE EXAMPLE( )

        IF( .TRUE. ) THEN
            CALL DO_SOMETHING()
        END IF ! ** .TRUE. **

    END
"""

with tempfile.NamedTemporaryFile('w', suffix=".f", delete=False) as file:
    file.write(FORTRAN)
crackfortran.crackfortran(file.name)
```

### Error message:

```shell
Reading fortran codes...
        Reading file '/var/folders/qv/js92y9x526sdmsmjdrylwkx80000gn/T/tmpmdqwsj7t.f' (format:fix,strict)
Traceback (most recent call last):
  File "/Users/lysnikolaou/repos/python/scipy/reexport.py", line 16, in <module>
    crackfortran.crackfortran(file.name)
  File "/Users/lysnikolaou/anaconda3/envs/scipy-dev/lib/python3.10/site-packages/numpy/f2py/crackfortran.py", line 3303, in crackfortran
    readfortrancode(files, crackline)
  File "/Users/lysnikolaou/anaconda3/envs/scipy-dev/lib/python3.10/site-packages/numpy/f2py/crackfortran.py", line 583, in readfortrancode
    dowithline(finalline)
  File "/Users/lysnikolaou/anaconda3/envs/scipy-dev/lib/python3.10/site-packages/numpy/f2py/crackfortran.py", line 822, in crackline
    raise Exception('crackline: groupcounter(=%s) is nonpositive. '
Exception: crackline: groupcounter(=0) is nonpositive. Check the blocks.
```

### Runtime information:

```python3
>>> import sys, numpy; print(numpy.__version__); print(sys.version)
1.24.2
3.10.10 | packaged by conda-forge | (main, Mar 24 2023, 20:17:34) [Clang 14.0.6 ]
>>> print(numpy.show_runtime())
[{'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3', 'SSE41', 'POPCNT', 'SSE42'],
                      'not_found': ['AVX',
                                    'F16C',
                                    'FMA3',
                                    'AVX2',
                                    'AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}},
 {'architecture': 'Nehalem',
  'filepath': '/Users/lysnikolaou/anaconda3/envs/scipy-dev/lib/libopenblasp-r0.3.21.dylib',
  'internal_api': 'openblas',
  'num_threads': 10,
  'prefix': 'libopenblas',
  'threading_layer': 'openmp',
  'user_api': 'blas',
  'version': '0.3.21'},
 {'filepath': '/Users/lysnikolaou/anaconda3/envs/scipy-dev/lib/libomp.dylib',
  'internal_api': 'openmp',
  'num_threads': 10,
  'prefix': 'libomp',
  'user_api': 'openmp',
  'version': None}]
```

### Context for the issue:

I'm currently working on scipy/scipy#14262, where I'm trying to regenerate scipy's Cython signatures for BLAS/LAPACK. The script that does this uses `f2py` to extract info from the BLAS/LAPACK fortran code. In four cases (`dbdsvdx`, `dtrevc3`, `sbdsvdx`, `strevc3`) `f2py` failed, because of this issue. For example, in `dbdsvdx`, the script fails due to [this line on LAPACK](https://github.com/Reference-LAPACK/lapack/blob/9f7abc2cc93fac3a5a8905307447c70163abdef2/SRC/dbdsvdx.f#LL697C37-L697C37).

### PR DESCRIPTION
**Pull request by HaoZeke**
Closes #23533.

### CODE REVIEW

File: numpy/f2py/crackfortran.py [hunk 1]
```diff
@@ -613,15 +613,15 @@ def readfortrancode(ffile, dowithline=show, istop=1):
 groupends = (r'end|endprogram|endblockdata|endmodule|endpythonmodule|'
              r'endinterface|endsubroutine|endfunction')
 endpattern = re.compile(
-    beforethisafter % ('', groupends, groupends, r'.*'), re.I), 'end'
+    beforethisafter % ('', groupends, groupends, '.*'), re.I), 'end'
 endifs = r'end\s*(if|do|where|select|while|forall|associate|block|' + \
          r'critical|enum|team)'
 endifpattern = re.compile(
-    beforethisafter % (r'[\w]*?', endifs, endifs, r'[\w\s]*'), re.I), 'endif'
+    beforethisafter % (r'[\w]*?', endifs, endifs, '.*'), re.I), 'endif'
```
Comment (HaoZeke) [inline numpy/f2py/crackfortran.py:620]: This is the fix, in keeping with the other pattern matches.

**Comment by HaoZeke** [comment]
Test failure on Azure seems to be unrelated:

```

tgt1       = array([ 5, 11])
tgt2       = array([[ 5, 11],
       [ 5, 11]])
v          = array([1.+0.j, 2.+0.j], dtype=complex64)
vec        = array([1, 2])

C:\hostedtoolcache\windows\Python\3.11.3\x86\Lib\site-packages\numpy\core\tests\test_multiarray.py:6925: RuntimeWarning
```

**Comment by charris** [comment]
> Test failure on Azure seems to be unrelated:

Yep, that error happens now and then on 32 bit windows.

**Comment by charris** [comment]
@HaoZeke Needs rebase.

**Comment by HaoZeke** [comment]
> @HaoZeke Needs rebase.

Should be good to go now 😊

**Comment by charris** [comment]
Thanks @HaoZeke . The `!` isn't really standard in F77, but I suspect all modern compilers support it.

**Pull request merged**